sycl/include/sycl/ext/oneapi/experimental/graph.hpp

```diff
@@ -130,22 +130,45 @@ struct graph_impl {
     MSchedule.clear();
   }
 
+  template <typename T>
+  node_ptr add(graph_ptr impl, T cgf, const std::vector<node_ptr> &dep = {}) {
+    node_ptr nodeImpl = std::make_shared<node_impl>(impl, cgf);
+    if (!dep.empty()) {
+      for (auto n : dep) {
+        n->register_successor(nodeImpl); // register successor
+        this->remove_root(nodeImpl);     // remove receiver from root node
+                                         // list
+      }
+    } else {
+      this->add_root(nodeImpl);
+    }
+    return nodeImpl;
+  }
+
   graph_impl() : MFirst(true) {}
 };
 
 } // namespace detail
 
-struct node {
-  detail::node_ptr MNode;
-  detail::graph_ptr MGraph;
-
+class node {
+public:
   template <typename T>
   node(detail::graph_ptr g, T cgf)
-      : MGraph(g), MNode(new detail::node_impl(g, cgf)){};
-  void register_successor(node n) { MNode->register_successor(n.MNode); }
-  void exec(sycl::queue q, sycl::event = sycl::event()) { MNode->exec(q); }
+      : MGraph(g), impl(new detail::node_impl(g, cgf)){};
```

Comment:
Another nitpick that's causing a compiler warning - can remove the trailing `;`

Reply:
Removed!